core/src/test/java/org/apache/calcite/rex/RexNormalizeTest.java

```diff
@@ -0,0 +1,118 @@
+    assertNodeEquals(node2, node3, true);
+
+    final RexNode node4 = eq(vVarchar(0), literal("01"));
+    final RexNode node5 = eq(literal("01"), vVarchar(0));
+    assertNodeEquals(node4, node5, true);
+  }
+
+  @Test void reversibleNormalizedToLess() {
+    // Same type operands.
+    final RexNode node0 = lt(vBool(0), vBool(0));
+    final RexNode node1 = gt(vBool(0), vBool(0));
+    assertNodeEquals(node0, node1, true);
+
+    final RexNode node2 = le(vBool(0), vBool(0));
+    final RexNode node3 = ge(vBool(0), vBool(0));
+    assertNodeEquals(node2, node3, true);
+
+    // Different type operands.
+    final RexNode node4 = lt(vSmallInt(0), vInt(1));
+    final RexNode node5 = gt(vInt(1), vSmallInt(0));
+    assertNodeEquals(node4, node5, true);
+
+    final RexNode node6 = le(vSmallInt(0), vInt(1));
+    final RexNode node7 = ge(vInt(1), vSmallInt(0));
+    assertNodeEquals(node6, node7, true);
+  }
+
+  @Test void reversibleDifferentArgTypesShouldNotBeShuffled() {
+    final RexNode node0 = plus(vSmallInt(1), vInt(0));
+    final RexNode node1 = plus(vInt(0), vSmallInt(1));
+    assertNodeEquals(node0, node1, false);
+
+    final RexNode node2 = mul(vSmallInt(0), vInt(1));
+    final RexNode node3 = mul(vInt(1), vSmallInt(0));
+    assertNodeEquals(node2, node3, false);
+  }
+
+  @Test void reversibleDifferentNullabilityArgsAreNormalized() {
+    final RexNode node0 = plus(vIntNotNull(0), vInt(1));
+    final RexNode node1 = plus(vInt(1), vIntNotNull(0));
+    assertNodeEquals(node0, node1, true);
+
+    final RexNode node2 = mul(vIntNotNull(1), vInt(0));
+    final RexNode node3 = mul(vInt(0), vIntNotNull(1));
+    assertNodeEquals(node2, node3, true);
+  }
+
+  @Test void symmetricalDifferentArgOps() {
+    final RexNode node0 = eq(vBool(0), vBool(1));
+    final RexNode node1 = eq(vBool(1), vBool(0));
+    assertNodeEquals(node0, node1, true);
+
+    final RexNode node2 = ne(vBool(0), vBool(1));
+    final RexNode node3 = ne(vBool(1), vBool(0));
+    assertNodeEquals(node2, node3, true);
+  }
+
+  @Test void reversibleDifferentArgOps() {
+    final RexNode node0 = lt(vBool(0), vBool(1));
+    final RexNode node1 = lt(vBool(1), vBool(0));
+    assertNodeEquals(node0, node1, false);
+
+    final RexNode node2 = le(vBool(0), vBool(1));
+    final RexNode node3 = le(vBool(1), vBool(0));
+    assertNodeEquals(node2, node3, false);
+
+    final RexNode node4 = gt(vBool(0), vBool(1));
+    final RexNode node5 = gt(vBool(1), vBool(0));
+    assertNodeEquals(node4, node5, false);
+
+    final RexNode node6 = ge(vBool(0), vBool(1));
+    final RexNode node7 = ge(vBool(1), vBool(0));
+    assertNodeEquals(node6, node7, false);
+  }
+
+  /** Assert whether two rex nodes are equal. */
+  private static void assertNodeEquals(RexNode node1, RexNode node2, boolean equal) {
+    StringBuilder reason = new StringBuilder("The two rex nodes expect to be ");
+    if (!equal) {
+      reason.append("not ");
+    }
+    reason.append("equal");
```

Comment:
@danny0405 , the reason must include the original node representations.

The error message like `The two rex nodes expect to be equal, expected: true got: false` does not really help to understand the nature of the failure.

Reply:
Well, let me fix it.